### ISSUE TITLE
watchDirectory option for watchOptions

### ISSUE THREAD

**Issue by sovetski** [webpack/webpack#17918]
## Feature request

**What is the expected behavior?**

To be able to add multiple custom directories to "watch" other than "assets".

**What is motivation or use case for adding/changing the behavior?**

I explained with more details here: https://github.com/symfony/webpack-encore/issues/1246

But it is also related to webpack, without using any external plugin which are not maintainted, like: https://github.com/Fridus/webpack-watch-files-plugin

**How should this be implemented in your opinion?**

```javascript
module.exports = {
  // other webpack configurations...

  watchOptions: {
    aggregateTimeout: 300, // Delay in milliseconds before rebuilding
    poll: 1000, // Check for changes every second
    // Specify the directory you want to watch
    // You can use an absolute or relative path
    watchDirectory: "customfoler/**/*.php",
  },
};
```

**Are you willing to work on this yourself?**
yes

**Comment by alexander-akait** [comment]
Why you need to watch `php` files by webpack? We have such options for `serve` (i.e. dev server to reload a page), but I don't see any reasons to just watch them by webpack itself

**Comment by sovetski** [comment]
> Why you need to watch `php` files by webpack? We have such options for `serve` (i.e. dev server to reload a page), but I don't see any reasons to just watch them by webpack itself

I put the link if you want to get more details: https://github.com/symfony/webpack-encore/issues/1246

But to explain here, when I use Symfony for example, and I use Bootstrap with "postcss-purgecss", it removes all unused classnames from the build, which means if I put for example "mt-5" classname on my php files (or twig), as it will not detect changes to "watch" and build, it will not include new classnames, so i need to update any js or css file to be able to build and get all used classnames. Let me know if it is not clear

**Comment by alexander-akait** [comment]
I see your problem, that tools parse php files and extract classes (loader/plugin)?

**Comment by sovetski** [comment]
> I see your problem, that tools parse php files and extract classes (loader/plugin)?

Exactly, it works also with PostCSS. Ex my `postcss.config.js` file content:

```javascript
module.exports = {
    plugins: [
        require("autoprefixer"),
        require("@fullhuman/postcss-purgecss")({
            content: [
                // PHP files from "src" and "vendor" directories that contain HTML/Twig templates
                // In "vendor" directory, bundles like "knplabs pagination" are using bootstrap classes
                "./src/**/*.php",
                "./vendor/**/*.php",
                "./vendor/**/*.html",
                "./vendor/**/*.twig",
                "./templates/**/*.twig",
                "./assets/**/*.js",
                "./assets/**/*.css",
            ],
            defaultExtractor: (content) => content.match(/[\w-/:]+(?<!:)/g) || [],
        }),
    ],
};`

The "watch" works as well with js,css but not with php files, default configuration from webpack I guess. Is there any possibility to add our custom files to watch?

**Comment by alexander-akait** [comment]
We have an API for such purposes and other tools use it (not only for webpack) - https://github.com/webpack-contrib/postcss-loader/blob/master/src/index.js#L191

`@fullhuman/postcss-purgecss` should use postcss message API, i.e. run this code:
```
result.messages.push(
  {
    type: "dependency",
    file: "/absolute/path/to/file.php",
    plugin: @fullhuman/postcss-purgecss,
  }
)
```

Webpack knows nothing about PHP files, also just watching PHP files and rerun compilation will very ineffective, using such API (example above) webpack rebuild only modules which depend `/absolute/path/to/file.php`, so you will have fast rebuild.

**Comment by alexander-akait** [comment]
Anyway, there is more old issue about it https://github.com/webpack/webpack/issues/13537, so let's close in favor https://github.com/webpack/webpack/issues/13537

**Comment by sovetski** [comment]
for the people who will come across this issue, my problem has not been solved at all

**Comment by duartecancela** [comment]
It works for me also. Thanks.
